### PR TITLE
Extend prop_list() to return properties across multiple lines

### PR DESCRIPTION
**Pull request by yegappan**

### CODE REVIEW

**Comment by brammool** [comment]
Thanks for working on this.
The "type" option currently only supports one type name.  The request was to use a pattern.  Maybe that is a bit slow (each type ID needs to be looked up to find the name to match with).  An alternative would be to pass a list of type names.

**Comment by vim-ml** [comment]
Hi Bram,

On Tue, Nov 16, 2021 at 8:51 AM Bram Moolenaar ***@***.***>
wrote:

> Thanks for working on this.
> The "type" option currently only supports one type name. The request was
> to use a pattern. Maybe that is a bit slow (each type ID needs to be looked
> up to find the name to match with). An alternative would be to pass a list
> of type names.
>
>
>
I have updated the PR to support specifying a list of property type names
and identifiers
to prop_list().

- Yegappan

**Comment by brammool** [comment]
Thanks for the update.  I'll leave this open for a couple of days to await any comments.

**Comment by bfrg** [comment]
How do we get all text-properties of a specific buffer?

In the examples you use `line('$')` but the function doesn't accept a buffer number. Currently the only way to obtain the number of lines in a buffer is through `getbufinfo(bufnr)[0].linecount` but this is a bit heavy considering how much data `getbufinfo()` could return. So wouldn't it be more efficient to accept `'$'` directly in `end_lnum`?

For example: `prop_list(1, {'bufnr': 4, 'end_lnum': '$', 'id': [123]})`.

**Comment by bfrg** [comment]
Hmm, Bram's comment didn't make it to github for some reasons. I thought emails were automatically mirrored.

I like the suggestion to add a new function `getline()` (or similar), since it could also be used in other places.

**Comment by vim-ml** [comment]
Hi,

On Thu, Nov 18, 2021 at 4:49 AM bfrg ***@***.***> wrote:

> How do we get all text-properties of a specific buffer?
>

Thanks for the feedback. I have updated the PR to use the last buffer line
number if
the end_lnum is greater than the last buffer line number. So you can use a
very large
value (9999999) for end_lnum.

- Yegappan


> In the examples you use line('$') but the function doesn't accept a
> buffer number. Currently the only way to obtain the number of lines in a
> buffer is through getbufinfo(bufnr)[0].linecount but this is a bit heavy
> considering how much data getbufinfo() could return. So wouldn't it be
> more efficient to accept '$' directly in end_lnum?
>
> For example: prop_list(1, {'bufnr': 4, 'end_lnum': '$', 'id': [123]}).
>
>
>

**Comment by puremourning** [comment]
Ping @bstaletic

File: src/textprop.c [hunk 1]
```diff
@@ -896,53 +896,258 @@ f_prop_find(typval_T *argvars, typval_T *rettv)
...
-							   sizeof(textprop_T));
-	    prop_fill_dict(d, &prop, buf);
-	    list_append_dict(rettv->vval.v_list, d);
+	    if (di->di_tv.vval.v_list == NULL)
+		return;
+
+	    prop_ids = get_prop_ids_from_list(di->di_tv.vval.v_list,
+		    &prop_ids_len);
+	    if (prop_ids == NULL)
+		return;
```
Comment (puremourning) [inline src/textprop.c]: does this leak prop_types? There seem to be a few early returns that might leak one of prop_types or prop_ids. Worth double-checking ?

**Comment by brammool** [comment]
The name of the item for the list of types is "type".  Now that it is a list it would be better to call this "types" or "typelist".
I think "types" is the best.

**Comment by vim-ml** [comment]
Hi,

On Fri, Nov 19, 2021 at 2:05 AM Ben Jackson ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/textprop.c
> <https://github.com/vim/vim/pull/9138#discussion_r753048618>:
>
> >
> -	    if (d == NULL)
> -		break;
> -	    mch_memmove(&prop, text + textlen + i * sizeof(textprop_T),
> -							   sizeof(textprop_T));
> -	    prop_fill_dict(d, &prop, buf);
> -	    list_append_dict(rettv->vval.v_list, d);
> +	    if (di->di_tv.vval.v_list == NULL)
> +		return;
> +
> +	    prop_ids = get_prop_ids_from_list(di->di_tv.vval.v_list,
> +		    &prop_ids_len);
> +	    if (prop_ids == NULL)
> +		return;
>
> does this leak prop_types? There seem to be a few early returns that might
> leak one of prop_types or prop_ids. Worth double-checking ?
>
>
> Yes. Thanks for catching this. The tests didn't detect this leak because
only
'types' or 'ids' item is used in the test.  I have updated the PR with a
fix for this
and added additional tests.

Regards,
Yegappan

**Comment by vim-ml** [comment]
Hi Bram,

On Fri, Nov 19, 2021 at 4:25 AM Bram Moolenaar
***@***.***> wrote:
>
> The name of the item for the list of types is "type". Now that it is a list it would be better to call this "types" or "typelist".
> I think "types" is the best.
>

I have updated the PR to rename "type" to "types" and "id" to "ids".

- Yegappan

**Pull request closed without merging**